### PR TITLE
Improve remote control API

### PR DESCRIPTION
**Pull request by SimonTeixidor**
With these commits it should be possible to write more feature complete remote controlling applications using librespot.

One issue I'm facing is that devices don't send their track list until they change state somehow. You can work around it by skipping a song and then going back, but I guess there must be a way to ask devices to send their track list. Perhaps you have any ideas?

### CODE REVIEW

File: src/spirc.rs [hunk 1]
```diff
@@ -164,22 +182,21 @@ impl SpircManager {
...
-            self.index = (self.index + 1) % self.tracks.len() as u32;
-            let track = self.tracks[self.index as usize];
-            self.player.load(track, true, 0);
-        } else {
-            self.notify_with_player_state(false, None, player_state);
+            if let Some(ref player) = self.player {
+                self.tracks.index = (self.tracks.index + 1) % self.tracks.ids.len() as u32;
+                let track = self.tracks.ids[self.tracks.index as usize];
+                player.load(track, true, 0);
+                return;
```
Comment (plietar) [inline src/spirc.rs]: Early returns like that are easy to miss.
Does it matter what happens in the no player case ?
Would that function ever be call anyway ?

File: src/spirc.rs [hunk 2]
```diff
@@ -57,7 +65,7 @@ impl SpircManager {
             name: name,
             ident: ident,
             device_type: 5,
-            can_play: true,
+            can_play: player_is_some,
```
Comment (plietar) [inline src/spirc.rs:68]: How do other devices see librespot when `can_play` is false ?
I remember it being grayed out but still in the list (ie locked phone).

I think you need to remove the `kCanBePlayer` capability from the `DeviceState` for it to not appear in the list.

File: src/util/spotify_id.rs [hunk 3]
```diff
@@ -88,3 +96,19 @@ impl FileId {
             .concat()
     }
 }
+
+#[test]
+fn test_base16() {
+    let str = "a719283ffb17abcd0192ea49b20139ff";
```
Comment (plietar) [inline src/util/spotify_id.rs]: Using `str` as a rust variable is very confusing

File: src/util/spotify_id.rs [hunk 4]
```diff
@@ -88,3 +96,19 @@ impl FileId {
             .concat()
     }
 }
+
+#[test]
```
Comment (plietar) [inline src/util/spotify_id.rs:100]: Nice 😄

File: src/spirc.rs (hunk 2 above)
Comment (SimonTeixidor) [reply src/spirc.rs:68]: Grayed out, saying "Unavailable for listening on". I removed the capability, but it didn't make any difference. We can keep it though, as it's probably the right thing to do.

Are you sure this isn't the intended behaviour? Are there devices you can buy which doesn't show up in the device list?

**Comment by SimonTeixidor** [comment]
I made some changes according to your comments :)

**Comment by SimonTeixidor** [comment]
I added one more commit, fixing the methods for modifying the queue and gave them more descriptive names.

**Comment by SimonTeixidor** [comment]
Is there anything stopping us from merging this? I'm happy to rebase :)

**Comment by sashahilton00** [comment]
@SimonPersson might be the tests included at the bottom of spotify_id.rs? Or perhaps Paul has just forgotten. Either way it would be nice to merge this since I've had to appropriate most of it into a local branch to work with it at the moment.

**Comment by plietar** [comment]
Sorry, I must have missed it. I'll have a more in depth look this weekend.

From my current understanding of the spotify connect protocol (which is not what is implemented), there should only be one list of track shared by all devices. Essentially every thing in `State` is shared, whereas `DeviceState` is device specific.

I think it would make sense for `SpircManager` to expose a unified API, so `tracks()`, `send_play()`, ... work, whether playback is happening locally or remotely.

That's outside the scope of this PR, but it's what I'd like to see `SpircManager` converge torwards. I tried to do this in [my big `v0.2` refactor](https://github.com/plietar/librespot/blob/v0.2/src/spirc/mod.rs)

**Comment by sashahilton00** [comment]
@plietar is this good to merge?

**Comment by plietar** [comment]
Looked a bit more, and I don't think this should be merged. I essentially outlined the reasons in my previous comment.
This PR diverges from the way Connect was designed and implemented by spotify. Admittedly, this is far from obvious, and the existing code isn't correct either.

I'll close this for now. Whenever `SpircManager` is fixed to do things properly these APIs can be added in a better way.
Thanks @SimonPersson any way for working on this

**Comment by sashahilton00** [comment]
Fair enough. Given that you have an idea as to how you want the library to be structured with regards to the Spotify protocol that is not immediately obvious from the code, are you planning to stop work on the current master and move to working on the v0.2 rewrite? The reason I ask is because I have begun writing a fairly lightweight extension for mopidy around the current librespot master, and am thus wondering if the library, or at least the interface to it, is going to change significantly to the point where I should just wait for the 0.2 rewrite to mature a little and use that for the extension?

**Comment by SimonTeixidor** [comment]
That's all right. If you could remember to ping me when there's been work towards a proper solution I'd appreciate it 👍

**Pull request closed without merging**